apps/state/lib/state/trip.ex

```diff
@@ -182,35 +162,45 @@ defmodule State.Trip do
     [new_trip | new_alternates]
   end
 
-  defp do_apply_filters(filters, trips \\ [])
+  defp do_apply_filters(%{routes: []}), do: []
+  defp do_apply_filters(%{route_patterns: []}), do: []
+  defp do_apply_filters(%{ids: []}), do: []
+  defp do_apply_filters(%{direction_id: _id} = filters) when map_size(filters) == 1, do: []
+  defp do_apply_filters(%{date: _date} = filters) when map_size(filters) == 1, do: []
 
-  defp do_apply_filters(%{routes: route_ids} = filters, trips) do
-    direction_id = filters[:direction_id]
-
-    filtered_trips =
-      for route_id <- route_ids do
-        by_route_and_direction_id(route_id, direction_id)
-      end
+  defp do_apply_filters(filters) do
+    matchers =
+      []
+      |> build_filters(:route_id, filters[:routes])
+      |> build_filters(:direction_id, filters[:direction_id])
+      |> build_filters(:route_pattern_id, filters[:route_patterns])
+      |> build_filters(:id, filters[:ids])
 
-    new_trips =
-      filtered_trips
-      |> List.flatten()
-      |> Enum.concat(trips)
+    trips = State.Trip.select(matchers) ++ State.Trip.Added.select(matchers)
```

Comment:
These will be slower if an index isn't provided; maybe `build_filters` should take/return an index as well?

Reply:
Not sure I fully understand how those indexes work but please take a look at the updated version and let me know if this is what is needed.